presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java

```diff
@@ -314,18 +314,20 @@ public void flushCache()
     private List<String> loadAllDatabases()
             throws Exception
     {
+        Optional<IMetaStoreClient> metastoreClient = Optional.of(clientProvider.createMetastoreClient());
         try {
             return retry()
                     .stopOnIllegalExceptions()
                     .run("getAllDatabases", stats.getGetAllDatabases().wrap(() -> {
-                        try (HiveMetastoreClient client = clientProvider.createMetastoreClient()) {
```

Comment:
Acquiring the client inside the retry handler is important: it allows connecting to a different metastore on failure.